Allow specifying multiple configurations within project manifests

kwridan opened the pull request:
Resolves https://github.com/tuist/tuist/issues/160

### Short description 📝

Xcode supports specifying multiple build configurations beyond the default "Debug" and "Release". Providing the ability to leverage multiple custom build configuration will allow Tuist's adoption by more projects.

### Solution 📦

Allow specifying multiple configurations from the project manifests. 

Note: The support and logic for dealing with multiple configurations has been in Tuist for some time as such it only required additional work to expose it in the Project manifests in a compatible fashion.

### Implementation 👩‍💻👨‍💻

> Detail in a checklist the steps that you took to implement the PR.

- [x] Expose multiple configurations within project manifest 
- [x]  Add fixture
- [x] Update scheme generator to look up appropriate build configuration
- [x] Add scheme linter to ensure it only references valid configurations
- [x] Update documentation
- [x] Update changelog 

### Test Plan 🛠

- Run `tuist generate` within `fixtures/ios_app_with_multi_configs`
- Verify the generated project contains the build configurations "Debug", "Beta" and "Release"
- Verify the generated project has a scheme "App-Beta" which uses the "Beta" configuration when performing the run action
- Verify that referencing a configuration within the scheme that isn't defined in the project flags a listing error 

### Questions / Discussion ❓

#### `CustomConfiguration`

To maintain backwards compatibility a new type was introduced to help associate a named build configuration (`BuildConfiguration`) with a `Configuration`. 

Multi config

```swift
let configurations: [CustomConfiguration] = [
    .debug(name: "Debug", xcconfig: "../ConfigurationFiles/Debug.xcconfig"),
    .release(name: "Beta", xcconfig: "../ConfigurationFiles/Beta.xcconfig"),
    .release(name: "Release", xcconfig: "../ConfigurationFiles/Release.xcconfig"),
]

let settings = Settings(configurations: configurations)
```

compatibility with previous debug / release: 

```swift
let debug = Configuration(xcconfig: "../ConfigurationFiles/Debug.xcconfig")
let release = Configuration(xcconfig: "../ConfigurationFiles/Release.xcconfig")

let settings = Settings(debug: debug, release: release)
```

This is the approach adopted by this PR, however there could be another we can follow.

Initially I attempted to update `Configuration` to include the name and variant however this would slightly change the behavior for existing users

```swift

public struct Configuration: Equatable, Codable {
    public let buildConfiguration: BuildConfiguration
    public let settings: [String: String]
    public let xcconfig: String?

    public init(buildConfiguration: BuildConfiguration = .debug, settings: [String: String] = [:], xcconfig: String? = nil) {
        self.buildConfiguration = buildConfiguration
        self.settings = settings
        self.xcconfig = xcconfig
    }

    public static func settings(_ settings: [String: String], xcconfig: String? = nil) -> Configuration {
        return Configuration(settings: settings, xcconfig: xcconfig)
    }
}

public extension Configuration {
    static func debug(name: String = "Debug", settings: [String: String] = [:], xcconfig: String? = nil) -> Configuration {
        return Configuration(buildConfiguration: .debug(name: name), settings: settings, xcconfig: xcconfig)
    }
    
    static func release(name: String = "Release", settings: [String: String] = [:], xcconfig: String? = nil) -> Configuration {
        return Configuration(buildConfiguration: .release(name: name), settings: settings, xcconfig: xcconfig)
    }
}
```

The change here would be any existing `Configuration` would need to get a default name and variant to maintain backwards compatibility, but that would result in 

```swift
let debug = Configuration(xcconfig: "../ConfigurationFiles/Debug.xcconfig")
let release = Configuration(xcconfig: "../ConfigurationFiles/Release.xcconfig") // debug under the hood due to defaults
let settings = Settings(debug: debug, release: release)
```

This could be mitigated internally in the `Settings` to re-create `Configuration` with the appropriate names and variants and only copying across any settings or xcconfig references.

```swift
public init(base: [String: String] = [:],
                debug: Configuration? = nil,
                release: Configuration? = nil,
                defaultSettings: DefaultSettings = .recommended) {
        configurations = [
            .debug(settings: debug?.settings ?? [:], xcconfig: debug?.xcconfig),
            .release(settings: release?.settings ?? [:], xcconfig: release?.xcconfig)
        ]
        self.base = base
        self.defaultSettings = defaultSettings
    }
```

The second approach reduces the types (and possibly any confusion that comes with it) however does open up some loopholes when using the default release / debug api. 

#### Scheme usage 

The scheme use can do with some evaluation: to ensure backwards compatibility the type `BuildConfiguration` does this work well at the point of use?

```swift
let betaScheme = Scheme(name: "App-Beta",
                         shared: true,
                         buildAction: BuildAction(targets: ["App"]),
                         runAction: RunAction(config: .release(name: "Beta"), executable: "App"))
```

vs. perhaps using a `String` to identify the configuration by name?

```swift
let betaScheme = Scheme(name: "App-Beta",
                         shared: true,
                         buildAction: BuildAction(targets: ["App"]),
                         runAction: RunAction(configurationName: "Beta", executable: "App"))
```

The linter needs to ensure a matching configuration exists either way.

pepibumur commented:
Thanks for exposing an API for this @kwridan. Answering some of your questions in the description of the PR:

> This is the approach adopted by this PR, however there could be another we can follow.

I lean towards your first approach:
- Fewer types keeps reduces the confusion as you pointed out.
- We optimize for the majority of users _(just a debug & release configuration)_ but allow extending it with extra configurations.

> vs. perhaps using a String to identify the configuration by name?

I'd go with defining the configuration using a string and having a linter that checks whether the configuration exists.

pepibumur commented on Sources/TuistKit/Generator/GeneratorModelLoader.swift:
```diff
@@ -285,15 +285,24 @@ extension TuistGenerator.InfoPlist.Value {
+
+    private static func buildConfigurationTuple(from customConfiguration: CustomConfiguration,
+                                                path: AbsolutePath) -> BuildConfigurationTuple {
+        let buildConfigruation = TuistGenerator.BuildConfiguration.from(manifest: customConfiguration.buildConfiguration)
```
Typo: buildConfigruation

pepibumur commented on Sources/TuistKit/Generator/GeneratorModelLoader.swift:
```diff
@@ -285,15 +285,24 @@ extension TuistGenerator.InfoPlist.Value {
 }
 
 extension TuistGenerator.Settings {
+    typealias BuildConfigurationTuple = (TuistGenerator.BuildConfiguration, TuistGenerator.Configuration?)
```
Looking at the code without having looked into this logic in a while, I find it confusing having to types named `BuildConfiguration` and `Configuration`. What's the difference? Could we use a different naming convention to clarify this?

kwridan replied:
Indeed hence why I was questioning adding yet a 3rd type 😀

- `BuildConfiguration` is the identifier of the configuration so to speak (e.g. "Beta"), the reason this is more elaborate than just a `String` is due to requiring the knowledge of the variant to apply the appropriate default settings.
- `Configuration` is the build settings + xcconfig file reference

ollieatkinson replied:
I agree, but I can't think of a better name. Maybe some header documentation on the types would lessen confusion when developing features inside of Tuist?

pepibumur commented on features/step_definitions/shared/xcode.rb:
```diff
@@ -22,3 +22,16 @@
+]
+
+  out, err, status = Open3.capture3("xcodebuild", *args)
+  search_for = "#{key} = #{value}"
```
If the underlying `xcodebuild` command fails, the error that we'd get with this logic complicates debugging the issue. I suggest adding the following:

```ruby
flunk(err) unless status.success?
```
That way we fail the test if the `xcodebuild` command fails.

pepibumur commented on Sources/TuistGenerator/Linter/SchemeLinter.swift:
```diff
@@ -0,0 +1,57 @@
+}
+
+private extension BuildConfiguration {
+    var linterDescription: String {
```
Do we need this considering that it's only being used in one place?

kwridan replied:
fair point, I will inline it - my thinking was seeing it's declared as private it's scope is local to this file.

davebcn87 commented on Sources/TuistGenerator/Generator/SchemesGenerator.swift:
```diff
@@ -437,4 +441,18 @@ final class SchemesGenerator: SchemesGenerating {
+        let debugConfiguration = project.settings.defaultDebugBuildConfiguration()
+        let buildConfiguration = debugConfiguration ?? project.settings.configurations.keys.first
+
+        return buildConfiguration?.name ?? "Debug"
```
just a minor thing... Can't we use `.debug` from `BuildConfiguration .Variant` instead of using the string directly? Just to prevent typos in the configuration in the future.

davebcn87 commented:
BTW, this looks awesome @kwridan! Really keen to see this merged and try it in my projects!

kwridan commented:
Pushed most of the changes - thanks for the feedback! 👍 

I went ahead with the alternate approach however pushed it to separate branch so we can compare the two approaches:

- I'll be calling the approach in this PR: _"Using a separate `CustomConfiguration` type"_
- vs the alternate on [this branch](https://github.com/bloomberg/tuist/tree/featutres/expose-multi-configs-alternative-api) - _"Reusing `Configuration`"_

Unless we make some breaking changes, going with Reusing `Configuration` has some funny cases:

- The following confusing case is allowed by the type system:

```swift
let settings = Settings(debug: .release(name: "Foo"), release: .debug(name: "Bar"))
```

The implementation on the branch ignores the names and variants passed into the `debug:` and `release:` parameters and always overrides them to debug and release regardless of was passed in.

- Additionally the reusing `Configuration` approach allows using the previous initializer that is left in for backwards compatibility

```swift

let settings = Settings(configurations: [
   Configuration(settings: [:], xcconfig: "path/foo.xcconfig"),
   Configuration(settings: [:], xcconfig: "path/bar.xcconfig")
])
```

This will produce two configurations with a default empty name which will get flagged by the linter on the branch.

Having played with both, I think there's some value in having a separate type, only at the manifest level to provide a clearer less error prone API. Alternatively we declare this a breaking change and require everyone to use the new initializers (which maybe too verbose 😞). 

Designing APIs is fun 🙂

pepibumur commented:
@kwridan I like the idea of having a separate type, but introducing breaking changes offers a poor experience to the users. What if we do the following:

1. Add a static method to `Settings` aligned with what you believe is a good API for Tuist. Let's say it's something along the lines of:
```swift
let settings: Settings = .settings([
  .debug(name: "Debug", settings: [:], xcconfig: "./Debug.xcconfig")
  .release(name: "Release", settings: [:], xcconfig: "./Release.xcconfig"),
  .release(name: "Beta", settings: [:], xcconfig: "./Beta.xcconfig")
])
```

2. Adjust the internals of `Settings` to deal with the two ways of initializing the settings. We can have an internal type that is not exposed to the user:

```swift
struct Settings {
  enum SettingsType {
    // We'll get rid of this one when we major-bump the version
    case deprecated(base: [String: String], debug: Configuration, release: Configuration)
    case configurations([SettingsConfiguration])
  }
}
```
3. Add a linter rule to that warns the user about the deprecation of the current approach and point them to the new approach.

I think that'd make the transition convenient for the users, but at the cost of complicating the internals. Since we'll clean up the code at some point, I wouldn't be too worried about the implementation being a bit more complex.

kwridan commented:
thanks all - should be ready for another look.

I believe we no longer need to deprecate settings type 🙂

- `BuildConfiguration` is now renamed to `PresetBuildConfiguration` to help clarify its purpose and allow its use in `Scheme` to maintain backwards compatibility without the odd cases.

- `CustomConfiguration` exposes:

```swift
.debug(name: String, settings: [String: String] = [:], xcconfig: String? = nil)
.release(name: String, settings: [String: String] = [:], xcconfig: String? = nil)
```

- `Settings` can be initialised with a list of `CustomConfiguration`s using

- `Settings` can still be initialised with `debug: Configuration` and `release: Configuration` - this is the default and is maintained for backwards compatibility and convenience to avoid unnecessary verbosity for the simple case of needing default settings and configurations.

- Added additional linting to flag duplicate configuration names

ollieatkinson commented on Sources/TuistGenerator/Models/Settings.swift:
```diff
@@ -55,9 +55,33 @@ public class Settings: Equatable {
+        return defaultConfiguration ?? debugConfigurations.first
+    }
+
+    /// Finds the default release `BuildConfiguration` if it exists, otherwise returns the first release configuration available.
```
Is this something which might catch people out if they are not careful? If order is important it might make sense to also surface some of this knowledge in the documentation.

ollieatkinson commented on Sources/TuistKit/Generator/ManifestLinter.swift:
```diff
@@ -0,0 +1,40 @@
+    func lint(project: Project) -> [LintingIssue]
+}
+
+class ManifestLinter: ManifestLinting {
```
I wonder if we need to have some linting at the Graph level which notifies the user of configurations not defined in manifests. I can imagine on larger projects it might be easy to miss one, or even worse, mis-spell one.

kwridan replied:
👍 
There's a scheme linter in place too to verify this

https://github.com/tuist/tuist/pull/451/files#diff-6a1a7d8fd087d39e1cf8d85558ece81c

kwridan replied:
Added https://github.com/tuist/tuist/issues/467 for a follow up

ollieatkinson commented on fixtures/ios_app_with_multi_configs/App/Project.swift:
```diff
@@ -0,0 +1,40 @@
+                         buildAction: BuildAction(targets: ["App"]),
+                         runAction: RunAction(configurationName: "Beta", executable: "App"))
+
+let project = Project(name: "MainApp",
```
During my testing, I added some configuration overrides for `Target` too. Which all works, might be a good idea to add them to this fixture to avoid regressions on that.

```swift
let project = Project(name: "MainApp",
                      settings: settings,
                      targets: [
                          Target(name: "App",
                                 platform: .iOS,
                                 product: .app,
                                 bundleId: "io.tuist.App",
                                 infoPlist: "Support/App-Info.plist",
                                 sources: "Sources/**",
                                 dependencies: [
                                    .project(target: "Framework1", path: "../Framework1"),
                                    .project(target: "Framework2", path: "../Framework2"),
                                ],
                                 settings: Settings(base: [
                                    "TARGET_BASE": "TARGET_BASE",
                                ], configurations: [
                                    .debug(name: "Debug", xcconfig: "../ConfigurationFiles/Target.Debug.xcconfig"),
                                    .release(name: "Release", xcconfig: "../ConfigurationFiles/Target.Release.xcconfig"),
                                ])
                          ),
                          Target(name: "AppTests",
                                 platform: .iOS,
                                 product: .unitTests,
                                 bundleId: "io.tuist.AppTests",
                                 infoPlist: "Support/AppTests-Info.plist",
                                 sources: "Tests/**",
                                 dependencies: [
                                     .target(name: "App"),
                          ]),
                      ], schemes: [betaScheme])
```

<img width="475" alt="image" src="https://user-images.githubusercontent.com/1382565/62042965-383bea80-b1f7-11e9-9e2f-ea96a4c2b53b.png">
